src/Surfnet/Stepup/Tests/Configuration/Value/AllowedSecondFactorListTest.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+/**
+ * Copyright 2017 SURFnet B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\Stepup\Tests\Configuration\Value;
+
+use PHPUnit_Framework_TestCase as TestCase;
+use Surfnet\Stepup\Configuration\Value\AllowedSecondFactorList;
+use Surfnet\StepupBundle\Value\SecondFactorType;
+
+class AllowedSecondFactorListTest extends TestCase
+{
+    /**
+     * @test
+     * @group domain
+     */
+    public function an_allowed_second_factor_list_deduplicates_its_second_factors_upon_creation()
+    {
+        $allowedSecondFactorList = AllowedSecondFactorList::ofTypes([
+            new SecondFactorType('sms'),
+            new SecondFactorType('sms'),
+        ]);
+
+        $this->assertCount(
+            1,
+            $allowedSecondFactorList,
+            'The AllowedSecondFactorList should have deduplicated its second factors, but it did not'
+        );
+    }
+
+    /**
+     * @test
+     * @group domain
+     *
+     * @dataProvider availableSecondFactorTypeProvider
+     * @param $availableSecondFactorType
+     */
+    public function a_blank_allowed_second_factor_list_allows_all_second_factors($availableSecondFactorType)
+    {
+        $allowedSecondFactorList = AllowedSecondFactorList::blank();
+
+        $isSecondFactorAllowed = $allowedSecondFactorList->allows($availableSecondFactorType);
+
+        $this->assertTrue(
+            $isSecondFactorAllowed,
+            'A blank allowed second factor list should allow all second factors but it does not'
```

Comment:
`sprintf('...does not allow SecondFactorType("%s")', $availableSecondFactorType)` -> identify what is wrong, not just that something that the failing test title already explains is wrong. However, since you've indexed the dataset with a proper key, you can ditch this message. (Test "a_blank.... failed with dataset [sms]" is indicative enough imo)